Cassandra 9306: cqlsh COPY tests

mambocab opened the pull request:
This adds a suite of tests for the COPY command in cqlsh, as described [here](https://issues.apache.org/jira/browse/CASSANDRA-9306).

Note that I've broken off the tests into their own module, as discussed in #188 -- this kept each test file a little more managably-sized and helped me organize utility functions. I can jam all the tests back into one file if anyone has strong feelings, but personally, I prefer them this way.

I've put off testing against a cluster with a failing node -- it may be better suited as part of [9172](https://issues.apache.org/jira/browse/CASSANDRA-9172), or part of the kitchen sink tests.

Requesting review from @thobbs, since he filed the Jira for these.

I'll squash before merging.

ptnapoleon commented:
You need to merge `master` into this branch and resolve the merge conflicts.

ptnapoleon commented:
Also you will definitely need to squash at least some of these commits away.

ptnapoleon commented on cqlsh_tests/cqlsh_tools.py:
```diff
@@ -0,0 +1,39 @@
+
+def csv_rows(filename, delimiter=None):
+    '''
+    Given a filename, opens a csv file and yeilds it line by line.
```
Typo in `yields`

ptnapoleon commented on cqlsh_tests/cqlsh_tools.py:
```diff
@@ -0,0 +1,39 @@
+            yield row
+
+
+def assert_csvs_items_equal(filename1, filename2):
```
Why `csvs` here?

mambocab replied:
You mean in the name `assert_csvs_items_equal`? Because it compares 2 csv files, so `csv`, but plural. Happy to change if you prefer a different name.

ptnapoleon replied:
The built in asserts use singular here, such as `assertListEqual`, but it's really a personal choice, so I'll leave it up to you.

ptnapoleon commented:
I'm sorry, but I think you need to comment these so as to satisfy `CONTRIBUTING.md`.

thobbs commented on cqlsh_tests/cqlsh_copy_tests.py:
```diff
@@ -0,0 +1,529 @@
+        write_rows_to_csv(tempfile.name, data)
+
+        self.node1.run_cqlsh(
+            """COPY ks.testquoted ("IdNumber", "select") FROM '{name}'""".format(name=tempfile.name))
```
Can you also test this without explicitly specifying the column names in the COPY command?

mambocab replied:
addressed in https://github.com/mambocab/cassandra-dtest/commit/70894809a37fb3a291f04712b38171e5d08c6747

thobbs commented on cqlsh_tests/cqlsh_copy_tests.py:
```diff
@@ -0,0 +1,529 @@
+
+        tempfile = NamedTemporaryFile()
+        self.node1.run_cqlsh(
+            """COPY ks.testquoted ("IdNumber", "select") TO '{name}'""".format(name=tempfile.name))
```
Ditto

thobbs commented on cqlsh_tests/cqlsh_copy_tests.py:
```diff
@@ -0,0 +1,529 @@
+    # are decoded differently. If this changes, un-skip these tests. In the
+    # meantime, the round-trip test for all datatypes shows that COPY works,
+    # though the contents of the CSV aren't defined in cqlshlib.
+    @unittest.skip('fails due to formatting differences in and out of cqlsh')
```
There are a couple of fixes for this I can think of.  One is to essentially do the same thing that cqlsh does and monkeypatch the driver during this test.  The other would be to process the CSV or CQL results to convert the blob columns to a matching format.

ptnapoleon commented:
Does nose pick up this file when run from the `cassandra-dtest/` directory?

ptnapoleon commented:
This seems like it could use the cluster reuse functionality, since you aren't changing cluster topology between tests. It would drastically improve run time.

mambocab commented:
> Does nose pick up these tests when run from the `cassandra-dtest/` directory?

Yep. You can verify with `nosetests --collect-only -v 2>&1 | grep cqlsh`.

> This seems like it could use the cluster reuse functionality

Can do. Working on a PR addressing comments now.

mambocab commented:
I believe I've addressed all comments. @ptnapoleon and @thobbs, can you re-review? Changes since initial push can be viewed [here](https://github.com/mambocab/cassandra-dtest/compare/dbfe0da...CASSANDRA-9306).

Again, I'll squash before commit.

ptnapoleon commented:
All my reservations are gone. Trusting you ran the tests here. +1. No comment on the changes Tyler wanted, I didn't check those out.

thobbs commented:
My comments all seem to be addressed as well, so +1.  Thanks!

mambocab commented:
Squashed into 2 commits so there's only one commit where `cqlsh_tests.py`'s content changes. Merging.
